lemmy_api/src/report.rs

```diff
@@ -0,0 +1,300 @@
+use actix_web::web::Data;
+use std::str::FromStr;
+
+use lemmy_db::{comment_report::*, comment_view::*, post_report::*, post_view::*, Reportable, ReportType,};
+use lemmy_structs::{blocking, report::*};
+use lemmy_utils::{APIError, ConnectionId, LemmyError};
+use lemmy_websocket::{LemmyContext, UserOperation, messages::SendUserRoomMessage};
+
+use crate::{check_community_ban, get_user_from_jwt, is_mod_or_admin, Perform};
+
+const MAX_REPORT_LEN: usize = 1000;
+
+#[async_trait::async_trait(?Send)]
+impl Perform for CreateReport {
+  type Response = CreateReportResponse;
+
+  async fn perform(
+    &self,
+    context: &Data<LemmyContext>,
+    websocket_id: Option<ConnectionId>,
+  ) -> Result<CreateReportResponse, LemmyError> {
+    let data: &CreateReport = &self;
+    let user = get_user_from_jwt(&data.auth, context.pool()).await?;
+
+    // check size of report and check for whitespace
+    let reason = data.reason.clone();
+    if reason.trim().is_empty() {
+      return Err(APIError::err("report_reason_required").into());
+    }
+    if reason.len() > MAX_REPORT_LEN {
+      return Err(APIError::err("report_too_long").into());
+    }
+
+    let report_type = ReportType::from_str(&data.report_type)?;
+    let user_id = user.id;
+    match report_type {
+      ReportType::Comment => { create_comment_report(context, data, user_id).await?; }
+      ReportType::Post => { create_post_report(context, data, user_id).await?; }
+    }
+
+    // to build on this, the user should get a success response, however
+    // mods should get a different response with more details
+    let res = CreateReportResponse { success: true };
+
+    context.chat_server().do_send(SendUserRoomMessage {
+      op: UserOperation::CreateReport,
+      response: res.clone(),
+      recipient_id: user.id,
+      websocket_id,
+    });
```

Comment:
Hrm... the recipients need to be the mods of this community, so we'll have to create the concept of `ModRoom`, similar to community_rooms:

`pub mod_rooms: HashMap<CommunityId, HashSet<ConnectionId>>,`

So then every mod can join and stay in these community mod rooms, and receive mod messages.

Reply:
added, i modeled it after the community one. not sure if that's where you want it though, so let me know ^_^

Reply:
also meant to ask, i don't need to do the websocket impl for these as far as who creates the report or for getting report lists, yeah? i assume those actions should just be done over http